crates/red_knot/src/symbols.rs

```diff
@@ -452,10 +502,16 @@ impl SymbolTableBuilder {
 impl PreorderVisitor<'_> for SymbolTableBuilder {
     fn visit_expr(&mut self, expr: &ast::Expr) {
         if let ast::Expr::Name(ast::ExprName { id, ctx, .. }) = expr {
-            self.add_symbol(id);
+            let flags = match ctx {
+                ast::ExprContext::Load => SymbolFlags::IS_USED,
+                ast::ExprContext::Store => SymbolFlags::IS_DEFINED,
+                ast::ExprContext::Del => SymbolFlags::IS_DEFINED,
+                ast::ExprContext::Invalid => SymbolFlags::empty(),
+            };
+            self.add_or_update_symbol(id, flags);
             if matches!(ctx, ast::ExprContext::Store | ast::ExprContext::Del) {
```

Comment:
Now that you're already matching on ctx above, this line might be simpler as a flags intersection check?

Reply:
Yes, I was wondering about this redundancy. Willfix.

Reply:
Added Copy+Clone on SymbolFlags:u8